lib/ts/AcidPtr.h

```diff
@@ -0,0 +1,198 @@
+/**
+  @file Extendible
+
+  @section license License
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+  @section details Details
+
+////////////////////////////////////////////
+  Implements advanced locking techniques:
+  * LockPool
+  * Writer_ptr
+*/
+
+#pragma once
+#include <mutex>
+#include <vector>
+#include <memory>
+
+//////////////////////////////////////////////////////////
+// Lock Pool
+/// Intended to make datasets thread safe by assigning locks to stripes of data, kind of like a bloom filter.
+/** Allocates a fixed number of locks and retrives one with a hash.
+ */
+template <typename Mutex_t> struct LockPool {
+  /**
+   * @param numLocks - use a prime number near the number of concurrent users you expect
+   */
+  LockPool(size_t num_locks) : mutexes(num_locks) {}
+
+  Mutex_t &
+  getMutex(size_t key_hash)
+  {
+    return mutexes[key_hash % size()];
+  }
+
+  size_t
+  size() const
+  {
+    return mutexes.size();
+  }
+
+  void
+  lockAll()
+  {
+    for (Mutex_t &m : mutexes) {
+      m.lock();
+    }
+  }
+
+  void
+  unlockAll()
+  {
+    for (Mutex_t &m : mutexes) {
+      m.unlock();
+    }
+  }
+
+private:
+  std::vector<Mutex_t> mutexes;
+
+  /// use the other constructor to define how many locks you want.
+  LockPool() = delete;
```

Comment:
I don't think this is useful - if you have a constructor defined, the compiler doesn't generate the default constructor. I'd be more tempted to `delete` the copy constructor.